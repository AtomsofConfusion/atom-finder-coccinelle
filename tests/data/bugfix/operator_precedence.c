#include <stdio.h>

int main() {
    int a = 4, b = 5, c = 2;
    int result;

    // Example 1: ! and &
    // Potentially Confusing
    result = !a & b;  // Might be read as (!a) & b
    printf("Result of !a & b: %d\n", result);

    // Clear
    result = !(a & b);  // Clearly negating the result of a & b
    printf("Result of !(a & b): %d\n", result);

    // Example 2: & and <<
    // Potentially Confusing
    result = a & b << c;  // Might be read as a & (b << c)
    printf("Result of a & b << c: %d\n", result);

    result = a << b & c; 
    // Clear
    result = (a & b) << c;  // Clear grouping of & and <<
    printf("Result of (a & b) << c: %d\n", result);

    // Example 3: ! and >>
    // Potentially Confusing
    result = !a >> b;  // Might be read as (!a) >> b
    printf("Result of !a >> b: %d\n", result);

    // Clear
    result = !(a >> b);  // Clearly negating the result of a >> b
    printf("Result of !(a >> b): %d\n", result);

    // Example 4: + and ?
    // Potentially Confusing
    result = a + b ? c : 100;  // Might be misread without grouping context
    printf("Result of a + b ? c : 100: %d\n", result);

    // Clear
    result = (a + b) ? c : 100;  // Clear intention of condition
    printf("Result of (a + b) ? c : 100: %d\n", result);

    // Example 5: | and ?
    // Potentially Confusing
    result = a | b ? c : 100;  // Could be confusing without parentheses
    printf("Result of a | b ? c : 100: %d\n", result);

    // Clear
    result = (a | b) ? c : 100;  // Explicit conditional expression
    printf("Result of (a | b) ? c : 100: %d\n", result);

    // Example 6: & and ?
    // Potentially Confusing
    result = a & b ? c : 100;  // Might be unclear without grouping
    printf("Result of a & b ? c : 100: %d\n", result);

    // Clear
    result = (a & b) ? c : 100;  // Clearly defined logical condition
    printf("Result of (a & b) ? c : 100: %d\n", result);

    // Example 7: ~ and |
    // Potentially Confusing
    result = ~a | b;  // Could be misread as negation applying to a or the expression
    printf("Result of ~a | b: %d\n", result);

    // Clear
    result = (~a) | b;  // Clearly applies negation to a first
    printf("Result of (~a) | b: %d\n", result);

    // Example 8: >> and *
    // Potentially Confusing
    result = a >> b * c;  // Might be read as a >> (b * c)
    printf("Result of a >> b * c: %d\n", result);

    // Clear
    result = (a >> b) * c;  // Explicit about shift first, then multiplication
    printf("Result of (a >> b) * c: %d\n", result);

    // Example 9: << and *
    // Potentially Confusing
    result = a << b * c;  // Might be read as a << (b * c)
    printf("Result of a << b * c: %d\n", result);

    result = a * b << c;  // Might be read as a << (b * c)

    // Clear
    result = (a << b) * c;  // Clear about order of operations
    printf("Result of (a << b) * c: %d\n", result);

    return 0;
}
