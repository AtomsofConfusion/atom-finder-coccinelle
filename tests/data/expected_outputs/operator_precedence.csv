operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,15,18,a + b * c
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,8,18,1 + 2 * 3
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,13,18,1 + 2 * ( 3 - 4 * 5 ) * 6
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,17,19,a * b + c
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,12,18,1 + 2 + 3 * 4
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,9,18,1 * 2 + 3
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,11,18,1 * 2 - 3
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,18,28,3 * b + 5
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,33,19,a + b << c
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,34,19,a + b >> c
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,35,19,a * b << c
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,36,19,a * b >> c
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,37,19,a & b + c
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,48,19,a & b > c
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,38,19,a + b & c
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,47,19,a < b & c
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,40,19,a + b ^ c
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,46,19,a & b ^ c
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,41,19,a | b + c
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,44,19,a | b & c
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,31,19,a | b ^ c
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,42,19,a + b | c
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,43,19,a & b | c
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,30,19,a ^ b | c
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,27,19,a + b && c
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,26,19,a - b && c
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,25,19,a * c && b
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,22,19,a % b && c
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,28,19,a << b && c
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,20,19,a | c && b
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,29,19,a ^ b && c
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,18,28,3 * b + 5 && c
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,18,19,a + 2 || 3 * b + 5 && c && 7
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,19,19,a & b || c
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,21,19,a ^ c || b
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,32,19,a && b || c
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,24,19,a && b * c
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,23,19,a && b % c
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,10,18,1 - 2 * 3
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,13,27,3 - 4 * 5
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,12,18,1 + 2 + 3 * 4 - 5 * 6
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,45,19,a ^ b & c
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,49,19,a || b & c
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,50,19,a || b ^ c
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,54,19,a && b | c
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,55,19,a && b + c
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,59,19,a && b + c
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,56,19,a && b - c
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,57,19,a && b << c
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,60,19,a && b >> c
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,51,19,a || b + c
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,52,19,a || b - c
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,53,19,a || b << c
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,61,19,a || b >> c
operator-precedence,/home/ubuntu/atom-finder-coccinelle/tests/data/inputs/operator_precedence.c,58,19,a || b && c

